### PR TITLE
WIP - do not merge. Please review my refactor for cloudwatch put_metric_data

### PR DESCRIPTION
**Pull request by chrisevett**
I refactored the way put_metric_data works so I can implement get_metric_statistics. I figured I'd get you guys' input on what I have in place already before I get started on the rest. 

This is babby's first python project so keep that in mind if you see anything that isnt idiomatic python. If it looks funny then it probably is 😆  🍼 

Some todos that I already know about:

Util.py needs tests
There are a bunch of commented out tests, these will be repurposed for get_metric_statistics and probably end up in their own file.
I have several test_fixture methods that will definitely change so its not so clunky 
The __str__ methods were for testing, i don't know if they should stay or not.

Anyway, please review and let me know if there are any major issues before I continue. Again, this refactor was to add support for all of the data in get_metric_statistics that didnt exist before.

### CODE REVIEW

File: moto/cloudwatch/models.py [hunk 1]
```diff
@@ -44,18 +60,127 @@ def __init__(self, namespace, name, value, dimensions):
...
+
+    def __init__(self, namespace, data):
+        self.namespace = namespace
+        self.data = data
+
+
+class MetricDatumv2(BaseModel):
+
+    def __init__(self, namespace, metricdata):
+        self.Namespace = namespace
```
Comment (JackDanger) [inline moto/cloudwatch/models.py]: It's unusual to have object attributes be uppercased like this. In this situation I'd expect `Namespace` to be a hardcoded class name. It makes me think this is how the code works:
```
class ANamespace:
    pass
class MetricDatumv2(BaseModel)
    def __init__(self):
        print(self.Namespace)
MetricDatumv2.Namespace = ANamespace
MetricDatumv2()  # prints 'ANamespace'
```

So I recommend using lowercase values for this and all similar variables throughout.

File: moto/cloudwatch/models.py [hunk 2]
```diff
@@ -44,18 +60,127 @@ def __init__(self, namespace, name, value, dimensions):
...
+    # parse the metricdata object
+
+        self.MetricData = [
```
Comment (JackDanger) [inline moto/cloudwatch/models.py]: Same here about the uppercase. `Metric` is great as a class name but I'd recommend `self.metric_data = Metric(...`

File: moto/cloudwatch/models.py [hunk 3]
```diff
@@ -44,18 +60,127 @@ def __init__(self, namespace, name, value, dimensions):
...
+            for metric in metricdata]
+
+    def __str__(self):
+        return "{} {}".format(self.Namespace, self.MetricData)
+
+
+class Metric():
+    def __init__(
+            self,
+            MetricName,
```
Comment (JackDanger) [inline moto/cloudwatch/models.py]: Same here for the variable names. In the boto3 api the keyword arguments to many of the API calls are camel cased but it's pretty unusual in Python and you won't find it in the Moto code base generally.

File: moto/cloudwatch/models.py [hunk 4]
```diff
@@ -102,10 +227,13 @@ def delete_alarms(self, alarm_names):
         for alarm_name in alarm_names:
             self.alarms.pop(alarm_name, None)
 
-    def put_metric_data(self, namespace, metric_data):
-        for name, value, dimensions in metric_data:
-            self.metric_data.append(MetricDatum(
-                namespace, name, value, dimensions))
+    def put_metric_data(self, Namespace, MetricData):
+        for datapoint in MetricData:
```
Comment (JackDanger) [inline moto/cloudwatch/models.py]: Do you need this instead?:
```
for datapoint in MetricData:
    self.metric_data.append(MetricDatumv2(Namespace, datapoint))
```

**Comment by JackDanger** [comment]
@chrisevett This is super good work. Thank you for tackling this! If you could make sure that old functionality still works so we don't go breaking somebody's test suite then your improvements are mergeable. I believe this means uncommenting the old tests and ensuring those code paths still work. This is great though and it's a huge help to everybody who uses AWS!

**Comment by chrisevett** [comment]
Thanks for the feed back. I will make the changes you requested and then send an update with the get_metric_statistics implementation included.

File: moto/cloudwatch/models.py (hunk 1 above)
Comment (thehesiod) [reply moto/cloudwatch/models.py]: ya it's against PEP, @JackDanger perhaps should enable some PEP flake tests?

File: moto/cloudwatch/models.py [hunk 5]
```diff
@@ -44,18 +60,127 @@ def __init__(self, namespace, name, value, dimensions):
...
+    def __init__(self, namespace, data):
+        self.namespace = namespace
+        self.data = data
+
+
+class MetricDatumv2(BaseModel):
+
+    def __init__(self, namespace, metricdata):
+        self.Namespace = namespace
+    # parse the metricdata object
```
[the remark this answers is not in the document]
Comment (thehesiod) [reply moto/cloudwatch/models.py]: another thing that can be caught by linter

File: moto/cloudwatch/models.py (hunk 1 above)
Comment (JackDanger) [reply moto/cloudwatch/models.py]: The test suite runs `flake8 ./moto` before every run and I'm not sure why this isn't being caught. Any ideas? Yes, we should absolutely tighten the flake8 constraints for developers. Maybe even adding autopep8 to help out would be useful.

Comment (thehesiod) [reply moto/cloudwatch/models.py]: just created https://github.com/spulec/moto/pull/1232

**Comment by thehesiod** [comment]
not sure how I didn't see this PR before I implemented something similar, @chrisevett could you check out what recently merged?

**Comment by chrisevett** [comment]
@thehesiod Sure thing. For some reason I thought we had merged this already but it looks like i was mistaken.

**Comment by chrisevett** [comment]
@JackDanger @thehesiod since this is so old I'm going to close this and open a new PR so I don't have to deal with funky merges/rebasing

**Pull request closed without merging**